### ISSUE TITLE
[jit] Support `named_parameters` in TorchScript

### ISSUE THREAD

**Issue by wanchaol** [pytorch/pytorch#46555]
We support it in C++ Script::Module 

https://github.com/pytorch/pytorch/blob/ccd9f3244babe972b87bd1f0ec43aade9ac54d49/torch/csrc/jit/api/module.h#L161

This is added in https://github.com/pytorch/pytorch/pull/28828

but it only works in the cpp_module (module._c) and didn't propagate well into python, I think it's because there're still some api difference with the [module.py's one](https://github.com/pytorch/pytorch/blob/f11f9a8c1f200c0798dd5043a958c563dfb00777/torch/nn/modules/module.py#L1092). 

We need to figure out a way to make it works in TorchScript

Example script:
```python
import torch


class TestModule(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.linear = torch.nn.Linear(2, 3)

    def forward(self, x):
        name_params = self.linear.named_parameters()
        for name, param in name_params:
            x += param

        return x


m = torch.jit.script(TestModule())
```

fails with 
```
Traceback (most recent call last):
  File "Desktop/test_script.py", line 17, in <module>
    m = torch.jit.script(TestModule())
  File "Desktop/test_script.py", line 7, in __init__
    self.linear = self.Linear(2, 3)
  File "/Users/wanchaol/Documents/pytorch/torch/nn/modules/module.py", line 778, in __getattr__
    raise ModuleAttributeError("'{}' object has no attribute '{}'".format(
torch.nn.modules.module.ModuleAttributeError: 'TestModule' object has no attribute 'Linear'
(ptdev) ➜  ~ python Desktop/test_script.py
Traceback (most recent call last):
  File "Desktop/test_script.py", line 17, in <module>
    m = torch.jit.script(TestModule())
  File "/Users/wanchaol/Documents/pytorch/torch/jit/_script.py", line 897, in script
    return torch.jit._recursive.create_script_module(
  File "/Users/wanchaol/Documents/pytorch/torch/jit/_recursive.py", line 352, in create_script_module
    return create_script_module_impl(nn_module, concrete_type, stubs_fn)
  File "/Users/wanchaol/Documents/pytorch/torch/jit/_recursive.py", line 410, in create_script_module_impl
    create_methods_and_properties_from_stubs(concrete_type, method_stubs, property_stubs)
  File "/Users/wanchaol/Documents/pytorch/torch/jit/_recursive.py", line 304, in create_methods_and_properties_from_stubs
    concrete_type._create_methods_and_properties(property_defs, property_rcbs, method_defs, method_rcbs, method_defaults)
  File "/Users/wanchaol/Documents/pytorch/torch/jit/_recursive.py", line 676, in compile_unbound_method
    stub = make_stub(fn, fn.__name__)
  File "/Users/wanchaol/Documents/pytorch/torch/jit/_recursive.py", line 37, in make_stub
    ast = get_jit_def(func, name, self_name="RecursiveScriptModule")
  File "/Users/wanchaol/Documents/pytorch/torch/jit/frontend.py", line 221, in get_jit_def
    return build_def(ctx, fn_def, type_line, def_name, self_name=self_name)
  File "/Users/wanchaol/Documents/pytorch/torch/jit/frontend.py", line 255, in build_def
    build_stmts(ctx, body))
  File "/Users/wanchaol/Documents/pytorch/torch/jit/frontend.py", line 127, in build_stmts
    stmts = [build_stmt(ctx, s) for s in stmts]
  File "/Users/wanchaol/Documents/pytorch/torch/jit/frontend.py", line 127, in <listcomp>
    stmts = [build_stmt(ctx, s) for s in stmts]
  File "/Users/wanchaol/Documents/pytorch/torch/jit/frontend.py", line 229, in __call__
    return method(ctx, node)
  File "/Users/wanchaol/Documents/pytorch/torch/jit/frontend.py", line 367, in build_Assign
    rhs = build_expr(ctx, stmt.value)
  File "/Users/wanchaol/Documents/pytorch/torch/jit/frontend.py", line 229, in __call__
    return method(ctx, node)
  File "/Users/wanchaol/Documents/pytorch/torch/jit/frontend.py", line 534, in build_Call
    args = [build_expr(ctx, py_arg) for py_arg in expr.args]
  File "/Users/wanchaol/Documents/pytorch/torch/jit/frontend.py", line 534, in <listcomp>
    args = [build_expr(ctx, py_arg) for py_arg in expr.args]
  File "/Users/wanchaol/Documents/pytorch/torch/jit/frontend.py", line 228, in __call__
    raise UnsupportedNodeError(ctx, node)
torch.jit.frontend.UnsupportedNodeError: Lambda aren't supported:
  File "/Users/wanchaol/Documents/pytorch/torch/nn/modules/module.py", line 1113
        """
        gen = self._named_members(
            lambda module: module._parameters.items(),
            ~ <--- HERE
            prefix=prefix, recurse=recurse)
        for elem in gen:
```

cc @ezyang @gchanan @zou3519 @bdhirsh @heitorschueroff @gmagogsfm

**Comment by wconstab** [comment]
@wanchaol to add a repro and more details about possible fix

**Comment by SplitInfinity** [comment]
I implemented PR #46784 for issue #46549, but we found a workaround for the use case (see issue for more details) so we decided not to land this. It would further entrench the assumption that a module can appear only once in a module hierarchy and make it difficult to fix (eventually, lol).
